src/app/replayer/replayer.ml

```diff
@@ -62,6 +64,42 @@ let create_ledger_as_list ledger =
   List.map accounts ~f:(fun acc ->
       Genesis_ledger_helper.Accounts.Single.of_account acc None )
 
+module First_pass_ledger_hashes = struct
+  (* ledger hashes after 1st pass, indexed by order of occurrence *)
+
+  module T = struct
+    type t = Ledger_hash.Stable.Latest.t * int
+    [@@deriving bin_io_unversioned, compare, sexp, hash]
+  end
+
+  include T
+  include Hashable.Make_binable (T)
+
+  let hash_set = Hash_set.create ()
+
+  let add =
+    let count = ref 0 in
+    fun ledger_hash ->
+      Base.Hash_set.add hash_set (ledger_hash, !count) ;
+      incr count
+
+  let find ledger_hash =
+    Base.Hash_set.find hash_set ~f:(fun (hash, _n) ->
```

Comment:
This is an `O(n)` operation, right?

If yes, I wonder why the hash set is used instead of simple list?

Reply:
Yes. Probably when I started on it, I thought we could do lookups on the `hash` alone.

The hash set is a mutable data structure, with the list we'd need to use a mutable reference. Not sure which is better.